src/colorizer/Collection.ts

```diff
@@ -307,25 +342,89 @@ export default class Collection {
+   * @param options optional configuration object containing any of the following properties:
+   *  - `fetchMethod` optional override for the fetch method.
+   *  - `reportWarning` optional callback for reporting warning messages during loading.
    * @throws an error if `url` is not a URL.
    * @returns a Promise of a new Collection object, either loaded from a collection JSON file or
    * generated as a wrapper around a single dataset.
    */
-  public static async loadFromAmbiguousUrl(url: string, fetchMethod = fetchWithTimeout): Promise<Collection> {
+  public static async loadFromAmbiguousUrl(
+    url: string,
+    options: Partial<{ fetchMethod: typeof fetchWithTimeout; reportWarning: ReportWarningCallback }> = {}
+  ): Promise<Collection> {
     // TODO: Also handle Nucmorph URLs that are pasted in? If website base URL matches, redirect?
 
     if (!isUrl(url)) {
       throw new Error(`Provided resource '${url}' is not a URL and cannot be loaded.`);
     }
 
+    let collectionLoadError: Error | null = null;
+    let datasetLoadError: Error | null = null;
     try {
-      return await Collection.loadCollection(url, fetchMethod);
+      return await Collection.loadCollection(url, options);
     } catch (e) {
+      collectionLoadError = e as Error;
+      console.warn(e);
       console.log("URL resource could not be parsed as a collection; attempting to make a single-database collection.");
     }
 
     // Could not load as a collection, attempt to load as a dataset.
-    return await Collection.makeCollectionFromSingleDataset(url);
+    try {
+      const collection = Collection.makeCollectionFromSingleDataset(url);
+      // Attempt to load the default dataset immediately to surface any loading errors.
+      const loadResult = await collection.tryLoadDataset(collection.getDefaultDatasetKey());
+      if (!loadResult.loaded) {
+        throw new Error(loadResult.errorMessage);
+      }
+      return collection;
+    } catch (e) {
+      datasetLoadError = e as Error;
+    }
+
+    // Assumption: if both loads failed and the file matches one of the default filenames,
+    // show only the relevant load error.
+    if (url.endsWith(DEFAULT_COLLECTION_FILENAME)) {
+      throw collectionLoadError;
+    } else if (url.endsWith(DEFAULT_DATASET_FILENAME)) {
+      throw datasetLoadError;
+    }
+
+    // Handle TypeError from failed fetch -> likely due to server being unreachable.
+    if (
+      collectionLoadError.message.startsWith(LoadErrorMessage.UNREACHABLE_COLLECTION) &&
+      datasetLoadError.message.includes(LoadErrorMessage.UNREACHABLE_MANIFEST)
+    ) {
+      throw new Error(
+        "Could not access either a collection or a dataset JSON at the provided URL." +
+          " This may be due to a network issue, the server being unreachable, or a misconfigured URL." +
+          " Please check your network access."
+      );
+    }
+
+    // Handle 404 errors from both collection and dataset fetches
+    if (
```

Comment:
should these be if else's ?

Reply:
They don't need to be since throwing an error stops execution? But I can see that being nicer for code readability